README.rst
===============
multivolumefile
===============

.. image:: https://coveralls.io/repos/github/miurahr/multivolume/badge.svg?branch=master
  :target: https://coveralls.io/github/miurahr/multivolume?branch=master

.. image:: https://github.com/miurahr/multivolume/workflows/Run%20Tox%20tests/badge.svg
  :target: https://github.com/miurahr/multivolume/actions

MultiVolumefile is a python library to provide a file-object wrapping multiple files
as virtually like as a single file. It inherit io.RawIOBase class and support some of
its standard methods.

See API details at `python library reference`_

.. _`python library reference`: https://docs.python.org/3/library/io.html

Status
======

The multivolumefile module is considered as ***Alpha*** stage.
All binary mode operation are implemented, but text mode is not completed, such as readline() method.
Because the library handles your files, it may damage your files.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

There is a good coverage on testing, but it may not be perfect.
You are welcome to add more test for a reliability.

Install
=======

You can install it as usual public libraries, you can use pip command

```
pip install multivolumefile
```

You are also able to add it to your setup.py/cfg as dependency.

Usages
------

- For reading multi-volume files, that has names `archive.7z.0001`, `archive.7z.0002` and so on,
  you can call multivolumefile as follows;

.. code-block::

    with multivolumefile.open('archive.7z', 'rb') as vol:
        data = vol.read(100)
        vol.seek(500)

- For writing multi-volue files, that has names `archive.7z.0001`, `archive.7z.0002` and so on,
  you can call multivolumefile as follows;


.. code-block::

    data = b'abcdefg'
    with multivolumefile.open('archive.7z', 'wb') as vol:
        size = vol.write(data)
        vol.seek(0)

you will see file `archive.7z.001` are written.

Limitations and known issues
============================

- Text mode is not implemented.
    - readlines(), writeline() and readline() raises NotImplementedError.
- fileno() is not supported and when call it, you will get RuntimeError exception.
- ***Caution***: When globbing existent volumes, it glob all files other than 4-digit extensions, it may break your data.


Contribution
============

You are welcome to contribute the project, as usual on github projects, Pull-Requests are always welcome.

License
=======

Multivolume is licensed under GNU Lesser General Public license version 2.1 or later.


Changelog.rst
=========
ChangeLog
=========

All notable changes to this project will be documented in this file.

`Unreleased`_
=============

Added
-----

Changed
-------

Fixed
-----

Deprecated
----------

Removed
-------

Security
--------

`v0.2.3`_
=========

Added
-----
* implement readall()

Chnaged
-------
* lint with black


`v0.2.2`_
=========

Added
-----

* Add py.typed file for type hinting.


`v0.2.1`_
=========

Added
-----

* Add `name` property that indicate basename of volumes
* Add `stat()` that return `stat_result` which has as mostly same methods as `os.stat_result` class
  except for platform dependent methods.


`v0.2.0`_
=========

Added
-----

* Type hint information bundled.

Fixed
-----

* Seek() returns current position.

Changed
-------

* Explanation of unsupported methods an modes in README

`v0.1.4`_
=========

Fixed
-----

* Fix append mode bug.

`v0.1.3`_
=========

Fixed
-----

* Fix added volume size become wrong

`v0.1.2`_
=========

Fixed
-----

* Fix append mode (#1)

`v0.1.1`_
=========

Fixed
-----

* Fin NotImplementedError when writing boudning of target files

`v0.1.0`_
=========

* ***API changed***

Added
-----

* Add mode 'x', 'xb' and 'xt'
* Add mode 'a', 'ab' and 'at'
* Support flush()

Changed
-------

* Change API: file argument of 'r' and 'rb' now single basename instead of list of files

`v0.0.5`_
=========

* Support context manager
* Support read functions.

.. History links
.. _Unreleased: https://github.com/miurahr/py7zr/compare/v0.2.2...HEAD
.. _v0.2.2: https://github.com/miurahr/py7zr/compare/v0.2.1...v0.2.2
.. _v0.2.1: https://github.com/miurahr/py7zr/compare/v0.2.0...v0.2.1
.. _v0.2.0: https://github.com/miurahr/py7zr/compare/v0.1.4...v0.2.0
.. _v0.1.4: https://github.com/miurahr/py7zr/compare/v0.1.3...v0.1.4
.. _v0.1.3: https://github.com/miurahr/py7zr/compare/v0.1.2...v0.1.3
.. _v0.1.2: https://github.com/miurahr/py7zr/compare/v0.1.1...v0.1.2
.. _v0.1.1: https://github.com/miurahr/py7zr/compare/v0.1.0...v0.1.1
.. _v0.1.0: https://github.com/miurahr/py7zr/compare/v0.0.5...v0.1.0
.. _v0.0.5: https://github.com/miurahr/py7zr/compare/v0.0.1...v0.0.5


tests/test_fuzzer.py
import io
import os
import pathlib
import sys

from hypothesis import given
from hypothesis import strategies as st
from pytest import fixture

import multivolumefile as MV

counter = 0


@fixture(scope="session")
def basepath(tmp_path_factory):
    return tmp_path_factory.getbasetemp()


def numbered_tempdir(base: pathlib.Path, common, suffix):
    global counter
    counter += 1
    tmpdir = base.joinpath(common).joinpath("{0}{1}".format(suffix, counter))
    os.makedirs(tmpdir, exist_ok=True)
    return tmpdir


@given(
    obj=st.binary(min_size=10, max_size=10 << 10),
    volume=st.integers(min_value=5, max_value=1 << 10),
)
def test_fuzzer_once(obj, volume, basepath):
    target = numbered_tempdir(basepath, "test_fuzzer", "once").joinpath("target.bin")
    with MV.open(target, mode="wb", volume=volume) as f:
        f.write(obj)
    with MV.open(target, mode="rb") as f:
        result = f.read()
    assert result == obj


@given(
    obj=st.binary(min_size=10, max_size=10 << 10),
    volume=st.integers(min_value=5, max_value=1 << 10),
    wblock=st.integers(min_value=5, max_value=5 << 9),
    rblock=st.integers(min_value=5, max_value=5 << 9),
)
def test_fuzzer_block(obj, volume, wblock, rblock, basepath):
    target = numbered_tempdir(basepath, "test_fuzzer", "block").joinpath("target.bin")
    src = io.BytesIO(obj)
    result = b""
    with MV.open(target, mode="wb", volume=volume) as f:
        data = src.read(wblock)
        while len(data) > 0:
            f.write(data)
            data = src.read(wblock)
    with MV.open(target, mode="rb") as f:
        data = f.read(rblock)
        while len(data) > 0:
            result += data
            data = f.read(rblock)
    assert result == obj


if __name__ == "__main__":
    import atheris  # type: ignore  # noqa

    atheris.Setup(sys.argv, test_fuzzer_block.hypothesis.fuzz_one_input)
    atheris.Fuzz()


tests/test_multivolume.py
import binascii
import hashlib
import os
import shutil

import pytest

import multivolumefile as MV

BLOCKSIZE = 8192
testdata_path = os.path.join(os.path.dirname(__file__), "data")
os.umask(0o022)


def test_read_check_sha1():
    target = os.path.join(testdata_path, "archive.7z")
    mv = MV.open(target, mode="rb")
    sha = hashlib.sha256()
    data = mv.read(BLOCKSIZE)
    while len(data) > 0:
        sha.update(data)
        data = mv.read(BLOCKSIZE)
    mv.close()
    assert sha.digest() == binascii.unhexlify(
        "e0cfeb8010d105d0e2d56d2f94d2b786ead352cef5bca579cb7689d08b3614d1"
    )


def test_read_seek():
    target = os.path.join(testdata_path, "archive.7z")
    mv = MV.open(target, mode="rb")
    assert mv.isatty() is False
    assert mv.seekable()
    mv.seek(40000)
    pos = mv.tell()
    assert pos == 40000
    b = mv.read(1)
    assert b == b'\x4d'
    mv.seek(-1001, os.SEEK_CUR)
    pos = mv.tell()
    assert pos == 39000
    mv.seek(-2337, os.SEEK_END)
    pos = mv.tell()
    assert pos == 50000
    mv.seek(15000, os.SEEK_SET)
    pos = mv.tell()
    assert pos == 15000
    b = mv.read(1)
    assert b == b"\x32"
    mv.close()
    assert mv.readable() is False
    mv.flush()


def test_read_context():
    target = os.path.join(testdata_path, "archive.7z")
    with MV.open(target, mode="rb") as mv:
        assert mv.readable()
        mv.seek(24900)
        data = mv.read(200)
        assert len(data) == 100
        data = mv.read(200)
        assert len(data) == 200


def test_readinto():
    target = os.path.join(testdata_path, "archive.7z")
    b = bytearray(200)
    with MV.open(target, mode="rb") as mv:
        mv.seek(24900)
        size = mv.readinto(b)
        assert size == 100


def test_read_boundary():
    target = os.path.join(testdata_path, "archive.7z")
    with MV.open(target, mode="rb") as mv:
        mv.seek(24999)
        b = mv.read(200)
        assert len(b) == 1


def test_readinto_boundary():
    target = os.path.join(testdata_path, "archive.7z")
    b = bytearray(200)
    with MV.open(target, mode="rb") as mv:
        mv.seek(25000)
        size = mv.readinto(b)
        assert size == 200
    sha = hashlib.sha256(b)
    assert (
        sha.digest()
        == b"6\xbc\x92\n\xa53tY\x97\xaa`!\xf3\xa7\xaa\xc3\\V\xac\xc1p\x97\xb0\xf5M\xc0)\x12\x0eD$\x9a"
    )


def test_readline():
    target = os.path.join(testdata_path, "archive.7z")
    with pytest.raises(NotImplementedError):
        with MV.open(target, mode="rb") as mv:
            mv.readline()


def test_readlines():
    target = os.path.join(testdata_path, "archive.7z")
    with pytest.raises(NotImplementedError):
        with MV.open(target, mode="rb") as mv:
            mv.readlines()


def test_readall():
    target = os.path.join(testdata_path, "archive.7z")
    with MV.open(target, mode="rb") as mv:
        b = mv.readall()
    sha = hashlib.sha256(b)
    assert sha.digest() == binascii.unhexlify(
        "e0cfeb8010d105d0e2d56d2f94d2b786ead352cef5bca579cb7689d08b3614d1"
    )


def test_write(tmp_path):
    target = tmp_path.joinpath("target.7z")
    with MV.open(target, mode="wb", volume=10240) as volume:
        assert volume.writable()
        with open(os.path.join(testdata_path, "archive.7z.001"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)

        with open(os.path.join(testdata_path, "archive.7z.002"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)
        assert volume.seekable()
        volume.seek(0)
        volume.seek(51000)
        volume.flush()
    created = tmp_path.joinpath("target.7z.0001")
    assert created.exists()
    assert created.stat().st_size == 10240


def test_write_default_volume(tmp_path):
    target = tmp_path.joinpath("target.7z").as_posix()
    with MV.open(target, mode="wb") as volume:
        with open(os.path.join(testdata_path, "archive.7z.001"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)
        with open(os.path.join(testdata_path, "archive.7z.002"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)
        assert volume.seekable()
        volume.seek(0)
        volume.seek(51000)
        volume.flush()
    created = tmp_path.joinpath("target.7z.0001")
    assert created.exists()
    assert created.stat().st_size == 52337


def test_write_posixpath(tmp_path):
    target = tmp_path.joinpath("target.7z").as_posix()
    with MV.open(target, mode="wb", volume=10240) as volume:
        assert volume.writable()
        with open(os.path.join(testdata_path, "archive.7z.001"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)

        with open(os.path.join(testdata_path, "archive.7z.002"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)
        assert volume.seekable()
        volume.seek(0)
        volume.seek(51000)
        volume.flush()
    created = tmp_path.joinpath("target.7z.0001")
    assert created.exists()
    assert created.stat().st_size == 10240


def test_write_boundary(tmp_path):
    target = tmp_path.joinpath("target.7z")
    with MV.open(target, mode="wb", volume=10240) as volume:
        assert volume.writable()
        with open(os.path.join(testdata_path, "archive.7z.001"), "rb") as r:
            data = r.read(10000)
            volume.write(data)
            data = r.read(250)
            volume.write(data)
        volume.flush()
    created = tmp_path.joinpath("target.7z.0002")
    assert created.exists()
    assert created.stat().st_size == 10


def test_write_exist(tmp_path):
    target = tmp_path.joinpath("target.7z")
    target_volume = tmp_path.joinpath("target.7z.0001")
    shutil.copyfile(os.path.join(testdata_path, "archive.7z.001"), target_volume)
    #
    with MV.open(target, mode="wb", volume=10240) as volume:
        assert volume.writable()
        with open(os.path.join(testdata_path, "archive.7z.001"), "rb") as r:
            data = r.read(10000)
            volume.write(data)
        volume.flush()
    created = tmp_path.joinpath("target.7z.0001")
    assert created.stat().st_size == 10000


def test_exclusive_write(tmp_path):
    target = tmp_path.joinpath("target.7z")
    with MV.open(target, mode="xb", volume=10240) as volume:
        assert volume.writable()
        with open(os.path.join(testdata_path, "archive.7z.001"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)

        with open(os.path.join(testdata_path, "archive.7z.002"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)
        assert volume.seekable()
        volume.seek(0)
        volume.seek(51000)
        volume.flush()
    created = tmp_path.joinpath("target.7z.0001")
    assert created.exists()
    assert created.stat().st_size == 10240


def test_exclusive_write_exist(tmp_path):
    target = tmp_path.joinpath("target.7z")
    target_volume = tmp_path.joinpath("target.7z.0001")
    shutil.copyfile(os.path.join(testdata_path, "archive.7z.001"), target_volume)
    with pytest.raises(FileExistsError):
        MV.open(target, "x")


def test_write_append1(tmp_path):
    target_volume = tmp_path.joinpath("target.7z.0001")
    with target_volume.open(mode="wb") as target:
        with open(os.path.join(testdata_path, "archive.7z.001"), "rb") as src:
            target.write(src.read(1000))
    #
    target = tmp_path.joinpath("target.7z")
    with MV.open(target, mode="ab", volume=10240) as volume:
        assert volume.seekable() is False
        with open(os.path.join(testdata_path, "archive.7z.001"), "rb") as src:
            src.seek(1000)
            volume.write(src.read(24000))
        with open(os.path.join(testdata_path, "archive.7z.002"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)
        volume.flush()
    #
    existed = tmp_path.joinpath("target.7z.0001")
    assert existed.exists()
    assert existed.stat().st_size == 10240
    created2 = tmp_path.joinpath("target.7z.0002")
    assert created2.exists()
    assert created2.stat().st_size == 10240
    created6 = tmp_path.joinpath("target.7z.0006")
    assert created6.exists()
    assert created6.stat().st_size == 1137


def test_write_append2(tmp_path):
    target = tmp_path.joinpath("target.7z")
    target_volume = tmp_path.joinpath("target.7z.0001")
    shutil.copyfile(os.path.join(testdata_path, "archive.7z.001"), target_volume)
    #
    with MV.open(target, mode="ab", volume=10240) as volume:
        with open(os.path.join(testdata_path, "archive.7z.002"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)
        volume.flush()
    #
    existed = tmp_path.joinpath("target.7z.0001")
    assert existed.exists()
    assert existed.stat().st_size == 25000
    created2 = tmp_path.joinpath("target.7z.0002")
    assert created2.exists()
    assert created2.stat().st_size == 10240
    created4 = tmp_path.joinpath("target.7z.0004")
    assert created4.exists()
    assert created4.stat().st_size == 6857  # 52337 - 25000 - 10240 * 2


def test_read_double_close():
    target = os.path.join(testdata_path, "archive.7z")
    mv = MV.open(target, mode="rb")
    assert not mv.closed
    mv.close()
    assert mv.closed
    mv.close()


def test_read_fileno():
    target = os.path.join(testdata_path, "archive.7z")
    mv = MV.open(target, mode="rb")
    with pytest.raises(RuntimeError):
        mv.fileno()
    mv.close()


def test_write_short_digits(tmp_path):
    target = tmp_path.joinpath("target.7z")
    with MV.MultiVolume(
        target, mode="wb", volume=10240, ext_digits=3, ext_start=0
    ) as volume:
        assert volume.writable()
        with open(os.path.join(testdata_path, "archive.7z.001"), "rb") as r:
            data = r.read(10000)
            volume.write(data)
            data = r.read(250)
            volume.write(data)
        volume.flush()
    created = tmp_path.joinpath("target.7z.001")
    assert created.exists()
    assert created.stat().st_size == 10


def test_write_hex_digits(tmp_path):
    target = tmp_path.joinpath("target.7z")
    with MV.MultiVolume(
        target, mode="wb", volume=800, hex=True, ext_digits=3, ext_start=0
    ) as volume:
        assert volume.writable()
        with open(os.path.join(testdata_path, "archive.7z.001"), "rb") as r:
            for _ in range(5):
                data = r.read(2000)
                volume.write(data)
            data = r.read(250)
            volume.write(data)
        volume.flush()
    created = tmp_path.joinpath("target.7z.00c")
    assert created.exists()
    assert created.stat().st_size == 650


def test_unsupported_open_mode(tmp_path):
    target = tmp_path.joinpath("target.7z")
    with pytest.raises(NotImplementedError):
        with MV.MultiVolume(target, mode="qb", volume=800):
            pass


def test_stat():
    target = os.path.join(testdata_path, "archive.7z")
    first_target = os.path.join(testdata_path, "archive.7z.001")
    with MV.open(target, mode="rb") as mv:
        st = mv.stat()
        assert st.st_size == 52337
        assert st.st_mtime == os.stat(first_target).st_mtime
        assert st.st_mtime_ns == os.stat(first_target).st_mtime_ns
        assert st.st_dev == os.stat(first_target).st_dev
        assert st.st_ino == 0



def test_truncate(tmp_path):
    target = tmp_path.joinpath("target.7z")
    target1 = tmp_path.joinpath("target.7z.0001")
    target2 = tmp_path.joinpath("target.7z.0002")
    target3 = tmp_path.joinpath("target.7z.0003")
    # write test data
    with MV.open(target, mode="wb", volume=10240) as volume:
        assert volume.writable()
        with open(os.path.join(testdata_path, "archive.7z.001"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)

        with open(os.path.join(testdata_path, "archive.7z.002"), "rb") as r:
            data = r.read(BLOCKSIZE)
            while len(data) > 0:
                volume.write(data)
                data = r.read(BLOCKSIZE)
        assert volume.seekable()
        volume.seek(0)
        volume.seek(51000)
        volume.flush()
        # check current status
        assert target1.stat().st_size == 10240
        assert target2.stat().st_size == 10240
        assert target3.exists()
        # run truncate
        assert volume.truncate(15000) == 15000
    # assert truncate
    assert target1.stat().st_size == 10240
    assert target2.stat().st_size == 4760
    assert not target3.exists()



multivolume.py
#!/usr/bin/env python
#
#    multi-volume file library
#    Copyright (C) 2020 Hiroshi Miura
#
#    This library is free software; you can redistribute it and/or
#    modify it under the terms of the GNU Lesser General Public
#    License as published by the Free Software Foundation; either
#    version 2.1 of the License, or (at your option) any later version.
#
#    This library is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#    Lesser General Public License for more details.
#
#    You should have received a copy of the GNU Lesser General Public
#    License along with this library; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
import io
import os
from typing import List, Union


def open(names, *, mode=None, volume=None) -> io.RawIOBase:
    return MultiVolume(names, mode=mode, volume=volume)


class MultiVolume(io.RawIOBase):

    def __init__(self, names, *, mode=None, volume=None):
        self._mode = mode
        self._closed = False
        self._files = []
        self._fileinfo = []
        self._position = 0
        self._positions = []
        if mode == 'rb' or mode == 'r':
            self._init_reader(names)
            pos = 0
            self._positions.append(pos)
            for size in self._fileinfo:
                pos += size
                self._positions.append(pos)
        elif mode == 'w':
            self._init_writer(names, mode, volume)
        else:
            raise NotImplementedError

    def _init_reader(self, names):
        for name in names:
            size = os.stat(name).st_size
            self._fileinfo.append(size)
            self._files.append(io.open(name, mode=self._mode))

    def _init_writer(self, names, mode, volume):
        raise NotImplementedError

    def _current_index(self):
        for i in range(len(self._positions) - 1):
            if self._positions[i] <= self._position < self._positions[i + 1]:
                pos = self._files[i].tell()
                offset = self._position - self._positions[i]
                if pos != offset:
                    self._files[i].seek(offset, io.SEEK_SET)
                return i
        if self._mode == 'rb' or self._mode == 'r':
            return len(self._files) - 1
        else:
            raise NotImplementedError

    def read(self, size: int = 1) -> bytes:
        current = self._current_index()
        file = self._files[current]
        data = file.read(size)
        if len(data) == 0 and current < len(self._files) - 1:
            current = self._current_index()
            file = self._files[current]
            file.seek(0)
            data += file.read(size)
        self._position += len(data)
        return data

    def readall(self) -> bytes:
        raise NotImplementedError

    def readinto(self, b: Union[bytes, bytearray, memoryview]):
        pass

    def write(self, b: Union[bytes, bytearray, memoryview]):
        raise NotImplementedError

    def close(self) -> None:
        self._closed = True
        for file in self._files:
            file.close()

    @property
    def closed(self) -> bool:
        return self._closed

    def fileno(self) -> int:
        """
        fileno() is incompatible with other implementation.
        multivolume handle multiple file object so we cannot return single value
        """
        return -1

    def flush(self) -> None:
        raise NotImplementedError

    def isatty(self) -> bool:
        return False

    def readable(self) -> bool:
        return True

    def readline(self, size: int = -1) -> bytes:
        raise NotImplementedError

    def readlines(self, hint: int = -1) -> List[bytes]:
        raise NotImplementedError

    def seek(self, offset, whence=io.SEEK_SET):
        if whence == io.SEEK_SET:
            target = offset
        elif whence == io.SEEK_CUR:
            target = self._position + offset
        else:
            target = self._positions[-1] + offset
        self._position = target
        i = len(self._files) - 1
        while i > 0 and target < self._positions[i]:
            i -= 1
        file = self._files[i]
        file.seek(target - self._positions[i], io.SEEK_SET)

    def seekable(self):
        return True

    def tell(self) -> int:
        return self._position

    def truncate(self, size=None):
        raise NotImplementedError

    def writable(self) -> bool:
        return False

    def writelines(self, lines):
        raise NotImplementedError

    def __del__(self):
        pass


multivolumefile/__init__.py
#!/usr/bin/env python
#
#    multi-volume file library
#    Copyright (C) 2020 Hiroshi Miura
#
#    This library is free software; you can redistribute it and/or
#    modify it under the terms of the GNU Lesser General Public
#    License as published by the Free Software Foundation; either
#    version 2.1 of the License, or (at your option) any later version.
#
#    This library is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#    Lesser General Public License for more details.
#
#    You should have received a copy of the GNU Lesser General Public
#    License along with this library; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
import contextlib
import io
import os
import pathlib
from mmap import mmap
from typing import Any, Container, List, Optional, Union

from .stat import stat_result

__all__ = ['stat_result', 'open', 'MultiVolume']

BLOCKSIZE = 16384


def open(name: Union[pathlib.Path, str], mode=None, volume=None) -> io.RawIOBase:
    return MultiVolume(name, mode=mode, volume=volume)


class _FileInfo:
    def __init__(self, filename, stat, size):
        self.filename = filename
        self.stat = stat
        self.size = size


class MultiVolume(io.RawIOBase, contextlib.AbstractContextManager):

    def __init__(self, basename: Union[pathlib.Path, str], mode: Optional[str] = 'r',
                 *, volume: Optional[int] = None, ext_digits: Optional[int] = 4, hex: Optional[bool] = False,
                 ext_start: Optional[int] = 1):
        self._mode = mode
        self._closed = False
        self._files = []  # type: List[object]
        self._fileinfo = []  # type: List[_FileInfo]
        self._position = 0
        self._positions = []
        self._digits = ext_digits
        self._start = ext_start
        self._hex = hex
        self.name = str(basename)
        if mode in ['rb', 'r', 'rt']:
            self._init_reader(basename)
        elif mode in ['wb', 'w', 'wt', 'xb', 'x', 'xt', 'ab', 'a', 'at']:
            if volume is None:
                self._volume_size = 10 * 1024 * 1024  # set default to 10MBytes
            else:
                self._volume_size = volume
            self._init_writer(basename)
        else:
            raise NotImplementedError

    def _glob_files(self, basename):
        if isinstance(basename, str):
            basename = pathlib.Path(basename)
        files = basename.parent.glob(basename.name + '.*')
        return sorted(files)

    def _init_reader(self, basename):
        pos = 0
        self._positions.append(pos)
        filenames = self._glob_files(basename)
        for name in filenames:
            stat = os.stat(name)
            size = os.stat(name).st_size
            self._fileinfo.append(_FileInfo(name, stat, size))
            self._files.append(io.open(name, mode=self._mode))
            pos += size
            self._positions.append(pos)

    def _init_writer(self, basename):
        if isinstance(basename, str):
            basename = pathlib.Path(basename)
        ext = '.{num:0{ext_digit}d}'.format(num=self._start, ext_digit=self._digits)
        target = basename.with_name(basename.name + ext)
        if target.exists():
            if self._mode in ['x', 'xb', 'xt']:
                raise FileExistsError
            elif self._mode in ['w', 'wb', 'wt']:
                file = io.open(target, mode=self._mode)
                self._files.append(file)
                file.truncate(0)
                stat = os.stat(target)
                self._fileinfo.append(_FileInfo(target, stat, self._volume_size))
                self._positions = [0, self._volume_size]
            elif self._mode in ['a', 'ab', 'at']:
                filenames = self._glob_files(basename)
                if self._mode == 'ab':
                    mode = 'rb'
                else:
                    mode = 'r'
                pos = 0
                size = 0
                self._positions = [0]
                for i in range(len(filenames)):
                    file = io.open(filenames[i], mode)
                    self._files.append(file)
                    stat = filenames[i].stat()
                    size = stat.st_size
                    self._fileinfo.append(_FileInfo(filenames[i], stat, size))
                    pos += size
                    self._positions.append(pos)
                    self._position = pos
                # last file
                if size >= self._volume_size:
                    self._add_volume()
                else:
                    self._files[-1].close()
                    self._files[-1] = io.open(filenames[len(filenames) - 1], mode=self._mode)
            else:
                raise NotImplementedError
        else:
            file = io.open(target, mode=self._mode)
            self._files.append(file)
            self._fileinfo.append(_FileInfo(target, os.stat(target), self._volume_size))
            self._positions = [0, self._volume_size]

    def _current_index(self):
        for i in range(len(self._positions) - 1):
            if self._positions[i] <= self._position < self._positions[i + 1]:
                pos = self._files[i].tell()
                offset = self._position - self._positions[i]
                if pos != offset:
                    self._files[i].seek(offset, io.SEEK_SET)
                return i
        return len(self._files) - 1

    def read(self, size: int = -1) -> bytes:
        if size == -1:
            return self.readall()
        current = self._current_index()
        file = self._files[current]
        data = file.read(size)
        self._position += len(data)
        return data

    def readall(self) -> bytes:
        result = b""
        data = self.read(BLOCKSIZE)
        while len(data) > 0:
            result += data
            data = self.read(BLOCKSIZE)
        return result

    def readinto(self, b: Union[bytearray, memoryview, Container[Any], mmap]) -> int:
        size = len(b)
        data = self.read(size)
        b[:len(data)] = data
        return len(data)

    def write(self, b: Union[bytes, bytearray, memoryview, Container[Any], mmap]) -> None:
        current = self._current_index()
        file = self._files[current]
        pos = file.tell()
        if pos + len(b) > self._volume_size:
            file.write(b[:self._volume_size - pos])
            self._position += self._volume_size - pos
            if current == len(self._files) - 1:
                self._add_volume()
            file = self._files[current + 1]
            file.seek(0)
            self.write(b[self._volume_size - pos:])  # recursive call
        else:
            file.write(b)
            self._position += len(b)

    def _add_volume(self):
        num = len(self._fileinfo) + self._start - 1
        if self._hex:
            last = self._fileinfo[-1].filename
            last_ext = '.{num:0{ext_digit}x}'.format(num=num, ext_digit=self._digits)
            assert last.suffix.endswith(last_ext)
            next_ext = '.{num:0{ext_digit}x}'.format(num=num + 1, ext_digit=self._digits)
            next = last.with_suffix(next_ext)
        else:
            last = self._fileinfo[-1].filename
            last_ext = '.{num:0{ext_digit}d}'.format(num=num, ext_digit=self._digits)
            assert last.suffix.endswith(last_ext)
            next_ext = '.{num:0{ext_digit}d}'.format(num=num + 1, ext_digit=self._digits)
            next = last.with_suffix(next_ext)
        self._files.append(io.open(next, self._mode))
        stat = os.stat(next)
        self._fileinfo.append(_FileInfo(next, stat, self._volume_size))
        pos = self._positions[-1]
        if pos != self._position:
            self._positions[-1] = self._position
        self._positions.append(self._positions[-1] + self._volume_size)

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        for file in self._files:
            file.close()

    @property
    def closed(self) -> bool:
        return self._closed

    def fileno(self) -> int:
        """
        fileno() is incompatible with other implementations.
        multivolume handle multiple file object so we cannot return single fd.
        """
        raise RuntimeError('fileno() is not supported.')

    def flush(self) -> None:
        if self._closed:
            return
        if self._mode == 'wb' or self._mode == 'w':
            for file in self._files:
                file.flush()

    def isatty(self) -> bool:
        return False

    def readable(self) -> bool:
        if self._closed:
            return False
        return all([f.readable() for f in self._files])

    def readline(self, size: Optional[int] = -1) -> bytes:
        raise NotImplementedError

    def readlines(self, hint: int = -1) -> List[bytes]:
        raise NotImplementedError

    def seek(self, offset: int, whence: Optional[int] = io.SEEK_SET) -> int:
        if whence == io.SEEK_SET:
            target = offset
        elif whence == io.SEEK_CUR:
            target = self._position + offset
        else:
            target = self._positions[-1] + offset
        self._position = target
        i = len(self._files) - 1
        while i > 0 and target < self._positions[i]:
            i -= 1
        file = self._files[i]
        file.seek(target - self._positions[i], io.SEEK_SET)
        return self._position

    def seekable(self) -> bool:
        if self._mode in ['ab', 'at', 'a']:
            return False
        else:
            return all([f.seekable() for f in self._files])

    def tell(self) -> int:
        return self._position

    def truncate(self, size: Optional[int] = None) -> int:
        raise NotImplementedError

    def writable(self) -> bool:
        return self._mode in ['wb', 'w', 'wt', 'x', 'xb', 'xt', 'ab', 'a', 'at']

    def writelines(self, lines):
        raise NotImplementedError

    def stat(self) -> stat_result:
        totalsize = 0
        for fi in self._fileinfo:
            totalsize += fi.size
        return stat_result(self._fileinfo[0].stat, totalsize)

    def __del__(self):
        # FIXME
        pass

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
